### ISSUE TITLE
`BufRead` requirement is paternalistic and impractical

### ISSUE THREAD

**Issue by TheBlueMatt** [rust-bitcoin/rust-bitcoin#3172]
While its true that reading rust-bitcoin types directly without a buffer is gonna be slow, sadly the Rust world uses the `Read` type everywhere, not `BufRead`. As a result, in many cases, you end up with a `Read` that you have to read bitcoin data out of...which you can't. The `BufRead` requirement was a cool idea in that it forced people to do a more efficient thing, but the less efficient thing isn't actually *wrong*, just slower, so should just be dropped IMO.

**Comment by apoelstra** [comment]
Can you give an example?

In a couple places we actually do use the `BufRead` trait to get extra efficiency (that we wouldn't get if we took a `Read` that was secretly a `BufRead`). So it'd be a little disappointing to lose this. Although it'd be just one of many many places where Rust's lack of specialization leads to inefficient code..

**Comment by apoelstra** [comment]
If you have a generic `R: Read` you already can't directly use rust-bitcoin functions because they take a `bitcoin::io::Read`. So you need to wrap stuff anyway.

Your point though about needing to buffer single bytes is well-taken .... if people need to wrap their stuff temporarily then they need to do this. (And scarily, if they attempt to temporarily wrap it using a "normal" `BufRead` then they may lose data, since the buffered reader will read data from the underlying reader but won't necessarily yield it before being dropped..)

**Comment by apoelstra** [comment]
> You wouldn't if we had a blanket implementation (which if we dropped the `BufRead` requirement we'd go back to - we have the blanket implementation on `std::BufRead` because without it trying to use `rust-bitcoin` is just too annoying).

We cannot have the blanket implementation because it will mean the `std` feature breaks any code which directly implemented `bitcoin::io::Read`. It has nothing to do with whether we use `Read` or `BufRead`. It's just Rust being broken.

**Comment by apoelstra** [comment]
I mean, everything about this is "Rust being broken" ... why the hell does the `Read` trait need std????

**Comment by apoelstra** [comment]
We have a blanket impl on the `BufReader` *type*, not on the `BufRead` *trait*. So there is no breakage there.

>Yes, users can use the library (slightly) wrong and end up getting smacked in the face for it.

If anybody implements `bitcoin::io::Read` on their types then their code will stop compiling as soon as somebody else impls the `std` feature on `bitcoin-io`. If they do this then they need to make sure that they are feature-gating their `bitcoin::io::Read` impl and converting it to a `std::io::Read` impl whenever the `std` feature of `bitcoin-io` is on. AFAIK this isn't possible to do (gating stuff on a feature-gate of another crate).

> That sucks, but also trying to use the library _without_ the blanket implementation is impossible

Yeah :(

**Comment by Kixunil** [comment]
> The Rust world uses `Read`, so you may use a method trait-bounded on `Read` and go to read a rust-bitcoin object as a part of a larger thing you're reading.

The *stream providers* use `Read` every other sane code uses `BufRead`. If you're bounding on `Read` you just need to push the responsibility to consumers.

> Right, so instead our goal of forcing people to "do the efficient thing" actually instead forced them to do a horrendously inefficient thing.

Or they could just bubble up the requirement, which should've been there from the beginning. I know it's a breakage and it sucks but I don't see any other solution. However it is true that people likely don't understand this (there would be no issue otherwise) so we need to work on the documentation.

> It's just Rust being broken.

I'm not sure, it seems unsolvable to me.

> why the hell does the std `Read` trait need std????

Because it's returning `std::io::Error` which had `last_os_error()` from the beginning and the situation would be somewhat salvageable by having a trait alias of a `core` trait with generic error type if they didn't decide to double down and add `read_vectored` which accepts an OS-specific type. So it will be in `std` forever and there's no solution.

> drop the `BufRead` stuff

We don't want to do that because of perf reasons. Also you still haven't provided any specific example where things go wrong, so I have no ways to help you and you have no way of convincing me there's a problem because all code I've seen so far has no problem and I can't even imagine what reasonable code would have a problem.

> `impl bitcoin_io::Read` for _every_ type in `std` which implements `std::io::Read` (probably easy to just slap a macro on it)

This was originally proposed but dropped as "too much work for little value" because one should normally use the buffered types which do have the impl. However if this helps you I'll be happy to ACK a PR that adds them. (But please, really check your code since you're likely doing something strange.)

**Comment by Kixunil** [comment]
And FTR on topic of "Rust being broken" a large part of motivation for `BufRead` is to workaround the fact that uninitialized reading is still unstable. But if it was stable I'd want to have uninitialized reading on `bitcoin-io::Read` too. :)

**Comment by Kixunil** [comment]
> And if someone has a library that already bounds on `Read`?

They aggain need to bubble up the requirement by changing it to `BufRead`. And the libraries depending on them too and so on until it reaches the exact place where putting the buffer is appropriate. This is how strong types work in general, so there's nothing weird here.

> but that doesn't mean we should have an unusable library.

rust-bitcoin is already unstable. By depending on it and exposing its types in the interface you're going to be unstable.

> As noted, the design here forces substantially _worse_ performance,

It only forces worse performance on badly-designed code and I wonder if the perf stays that bad after inlining and other optimizations.



> > The stream providers use Read every other sane code uses BufRead. If you're bounding on Read you just need to push the responsibility to consumers.
> 
> And if someone has a library that already bounds on `Read`? I'm sorry you disagree with what others do, but that doesn't mean we should have an unusable library.
> 
> > We don't want to do that because of perf reasons. Also you still haven't provided any specific example where things go wrong, so I have no ways to help you and you have no way of convincing me there's a problem because all code I've seen so far has no problem and I can't even imagine what reasonable code would have a problem.
 
> As noted, the design here forces substantially _worse_ performance, not to mention wrappers in downstream codebases which also sucks.

You already have to use `BufReader` if your type is not buffered - you get super slow code otherwise. Complaining about having to write one line of code that you should've had anyway makes no sense.

**Comment by Kixunil** [comment]
> _everything_ is bound on `Read`

It's just painted into a corner but doesn't matter as long as it cannot benefit from optimizations and doesn't pass the reader to some lower-level unstable library.

> I said un_usable_ not un_stable_.

Sorry about misreading. My point is it's usable but unstable if you pass down the requirement.

That all being said, I have an idea how we could fix this - by doing exactly what Rust should've done (I think):

```rust
pub trait Read {
    type Buffered: BufRead = NoReader; // does our MSRV support defaults here?
    type Unbuffered: Read = Self;

    fn as_buffered(&mut self) -> Result<&mut Self::Buffered, &mut Self::Unbuffered> {
        Err(self)
    }

    // remaining methods
}

// all buffered types get this impl
impl Read for &'_ [u8] {
    type Buffered = Self;
    type Unbuffered = NoReader;

    fn as_buffered(&mut self) -> Result<&mut Self::Buffered, &mut Self::Unbuffered> {
        Ok(self)
    }
}

pub enum NoReader {}

impl Read for NoReader { /* ... matches on self */ }
impl Read for NoReader { /* ... matches on self */ }

fn decode(reader: &mut impl Read) -> io::Result<Foo> {
    match reader.as_buffered() {
        Ok(reader) => decode_fast(reader),
        Err(reader) => decode_slow(reader),
    }
}
```

Note however that the slow path will be really horrible in some cases - reading stuff byte-by-byte. E.g. `push_decode` requires a slice of `u8`. But honestly, I don't care about bad performance of people who won't listen to my warnings.

Also this makes trait objects more annoying but we can add helpers to work around that.

**Comment by Kixunil** [comment]
> This works great, as long as you can pass down the requirement.

**WHO** can't? Give me one example. I've asked for examples at least two times and you didn't gave any. Are you doing this on purpose?!

**Comment by Kixunil** [comment]
> Maybe for 0.32 we do #3173 and then we can do something smarter later.

Then we will have to change it again and it will just piss off completely everyone. Just do the right change from the beginning.

**Comment by apoelstra** [comment]
serde_json::from_reader uses `Read`, as an external example.

rust-elements uses `Read` (and we are going to just reimplement `Encodable`/`Decodable` on the rust-bitcoin types we need to avoid bubbling up the bound -- though there we are also trying to avoid exposing `bitcoin::io` types entirely because we haven't done the work to make this useful instead of just anonying).

**Comment by apoelstra** [comment]
I agree with @TheBlueMatt that there's no way we are going to move the entire ecosystem from `Read` to `BufRead`, and we *definitely* aren't going to do so starting from a crate designed to help nostd bitcoin users, which is pretty-much as niche as you can get.

Having said this, @Kixunil your proposal where we switch on bufferedness or not sounds good to me.

I also would really like to get the blanket impl for `std::io::Read` to work.... I wonder if there is some combination of sealing or auxiliary traits that could do that.

**Comment by Kixunil** [comment]
> but we do have lots of users reading from various streams and `BufRead` isn't always practical

Which specific one would have a problem? Do they have horrible performance right now?

> restrict our users arbitrarily

It's not arbitrary if the upstream crate requires it.

> inside other messages/serialization libraries

I've never seen a library that calls back with `Read` objects rather than `&[u8]` which already implements `BufRead`. Can you name one?

> so it shouldn't upset anyone (compared to what already exists)?

It will upset people who started upgrading to 0.32 now, then abort changing the trait bounds and then learn they have to do the change anyway. And it will also annoy myself and likely other maintainers for having to do bunch of useless work.

> serde_json::from_reader uses `Read`, as an external example.

It doesn't pass it to anything so there's no breakage. (Though they did paint themselves into the corner.)

> rust-elements uses `Read` (and we are going to just reimplement `Encodable`/`Decodable` on the rust-bitcoin types we need to avoid bubbling up the bound -- though there we are also trying to avoid exposing `bitcoin::io` types entirely because we haven't done the work to make this useful instead of just anonying).

Does it pass around the bound though? A link to code would be nice for checking.

Note also that I've remembered that unbuffered reading doesn't make `push_decode` just unperformant but also entirely broken when reading empty types (well, TBH that's a niche case). But we can improve both by adding a method `min_required_bytes_hint(&self) -> usize` to `Decoder`. It'll make decoders much more complex but at least one can first create a zeroed buffer, then in a loop copy chunks of bytes to it rather than individual byte and pass them to the `bytes_received` method.

**Comment by apoelstra** [comment]
> Does it pass around the bound though? A link to code would be nice for checking.

Yes. https://github.com/ElementsProject/rust-elements/blob/6606a0da18cd17709274da70f3f3c69beb3b78e9/src/encode.rs#L213-L217/

**Comment by apoelstra** [comment]
Because it's obnoxious to my users and provides no benefit because the crate doesn't work with nostd anyway (and its primary users, the Liquid functionaries and Esplora, don't need it to be, so it's not a high priority).

"Not stable" doesn't mean that I should force API changes on people.

**Comment by apoelstra** [comment]
For sure, we can (and may, once it's useful, especially if we have `primitives` out and we can start reducing tech debt in `rust-elements`) change `std::io::Read` to `std::io::BufRead`.

The problem with rust-elements is actually separate/more general than the problem this issue is about. In our case we're annoyed about having to switch from `std::io` types to `bitcoin::io` types (which would be resolved if we could have the blanket impl somehow..) and the buffered vs non-buffered thing is really irrelevant.

**Comment by Kixunil** [comment]
Oh crap, I've noticed we're also missing `std` bridging types. :man_facepalming: #3175

**Comment by apoelstra** [comment]
@TheBlueMatt well, the issue is that if we take a `Read` we *can't* "trivially do the efficient thing". We can trivially do something more efficient than a fake-buffer reader, but it's impossible for us to get the efficiency that directly taking a `BufRead` would get us.

I think maybe Kix was right in one of his above proposals that we should have separate decoding methods depending on whether you have a buffered reader (with a default impl where the one that takes a buffered reader just calls `Read` methods on it, but which can be overridden to use `BufRead` methods if that will help).

**Comment by apoelstra** [comment]
It's not. `Read::read` copies data out of the buffer into the `&mut [u8]` you pass into it. `BufRead::fill_buf` returns a reference to the internal buffer without any copying.

**Comment by Kixunil** [comment]
On top of what @apoelstra said, the `BufReader` type is able to use the nightly version of `std::io::Read` which provides a way to avoid zeroing the buffer. Using the `read` method forces you to zero the buffer (no, passing `unsafe { core::mem::uninitialized() }` is not sound). So in the worst case you zero out as much data as you read. And once the method stabilizes it'll still be helpful to not have to wait two years.

Also I hypothesize that it can make fuzzing faster since fuzzers give us `&[u8]` which we can read directly and the `fill_buf` method just returns the same slice so after inlining and other optimizations the compiler may produce slightly more optimal code. So we get more fuzz runs per second.

**Comment by Kixunil** [comment]
> though last I checked it sadly does sometimes :(

Well, if it didn't the Rust devs wouldn't have added the uninitialized buffer API. :)

**Comment by Kixunil** [comment]
Oh, that's true. Though I'd be surprised if it can optimize things like reading to an uninitailized part of `Vec`.

**Comment by Kixunil** [comment]
FYI I've figured out a reasonable solution for `push_decode`: https://github.com/Kixunil/push_decode/commit/d49a109acf6698de3a8ad1d5ded3e573373f6cf4

Though this will force us to write more code in `bitcoin` and the operations are likely *still slower* than when using `BufRead`.

**Comment by apoelstra** [comment]
Nice! Yes, this may be slower because of the 0-initializing and the copying between extra buffers. I think we may have to live with it. And anyway it's only in a separate function which is explicitly marked as being slow so that's fine.

(BTW the doccomment there is another place where I'd like you to tone down the language before merging :) or at least, if you're going to complain about "poorly-designed APIs" at least be clear that it's the whole Rust ecosystem at fault, not specific unnamed developers).

**Comment by Kixunil** [comment]
> at least be clear that it's the whole Rust ecosystem at fault

I haven't implied anything. I'm not convinced it's even *most* Rust ecosystem fault but it's quite possible it is. And for instance, I design my APIs with `BufRead` in mind where it matters. :)

**Comment by Kixunil** [comment]
I've been thinking about this and came to conclusion that this issue is vastly overblown. (But yet, I'm working on a helper.)

If we analyze the problem, we have two kinds of breakages:

* A library depends on `bitcoin` - we discussed this above, it's just inherently unstable situation and breakages should be expected. We're not responsible for someone being unable to update their API - we already had a ton of breakages and we will have more. Just bubble up the bound and call it a day.
* A library provides a reader that doesn't implement `BufRead` - let's analyze this.

It was suggested that `Read` would be provided by a deserialization format. The only way that I can think of that makes it seemingly impossible to use is something like this:

```rust
trait Visitor {
    fn visit_bytes<R: Read>(reader: R) -> io::Result<Self>;
}
```

A consumer of `bitcoin` is also the implementor of `Visitor` and needs to deserialize the bytes in the method.

If the deserialization library is in any way sane it must have a mechanism to know when the reading of the field stops. IOW attempting to read past the value should not produce a bug! This means that just wrapping the `reader` in `BufReader` is sufficient to do the right thing. There's no need to even do one-byte buffering. It might be inefficient in copying the buffer twice if there's already a buffer but that could be solved backward-compatibly if the maintainer adds a method to `Visitor` accepting a `BufRead`.

**Comment by Kixunil** [comment]
> This seems like another case of #3166 - API breaks from rust-bitcoin have cost downstream projects a massive amount of energy

Sadly, this is going to happen for a while. There's not much we can do unless you have a time machine to prevent the old APIs to be in the crate in the first place. :(

> You assume a _lot_ about serialization libraries.

Only things that are clearly required. It's impossible to implement it otherwise without breaking layers. I know of no format that does it. Do you have any counter example?

**Comment by apoelstra** [comment]
>Sadly, this is going to happen for a while. There's not much we can do unless you have a time machine to prevent the old APIs to be in the crate in the first place. :(

Yes but here we have a clear opportunity to make @TheBlueMatt (and others') life easier with the currently-released version of rust-bitcoin. With push-based decoding we will have more room to iterate to get a broadly-usable API since we won't have major libraries pledged to stick with an already-released version.

In addition I think that, unstable or no, it's unreasonable for us to say "everything requires `BufRead`, just propagate it outward indefinitely". IMO it's fine if we have a second-class API for `Read` users, even a slow one, but we need something to exist.

**Comment by stevenroose** [comment]
I didn't follow the entire discussion, but the `read_to_limit` method on `bitcoin::io::Read` is quite pointless. Used only once, in a psbt impl and can quite easily be converted into something on top of `std::io::Read`.

I was trying to figure out why we had `bitcoin::io::Read` in the first place.

**Comment by apoelstra** [comment]
It's because we already had `read_to_end` (copied from `std`) but since this requires extra work to use safely, we did that work for the user by providing the extra `read_to_limit` method, which as a trait method with a default implementation, has basically zero cost to users.

**Comment by Kixunil** [comment]
@stevenroose LOOOOOL!!!! This gave me a huge laugh because it's related to the exciting news I have:

**I may have found a solution to get (most of?) the benefits of `BufRead` without requiring it!**

The thing is that I believe the only operations which we can sensibly optimized are reading of large byte vectors and large byte arrays.

And we have a method for the former which can be specialized by overriding it. ;) So @stevenroose there's your reason to use it and a good reason to exist. I propose that we also add a `read_byte_array<const N: usize>(&mut self) -> io::Result<[u8; N]>` that will read into an uninitialized buffer in overridden `BufRead` implementations (we provide tools to do that sanely). The `std` implementations will just override `read_to_limit` by calling `self.inner.by_ref().take(limit).read_to_end().map_err(Into::into)` and leave `read_byte_array` as-is until buf reading stabilizes in `std`.

I was also thinking of implementing some kind of linting to prevent people from forgetting to override those methods when implementing `BufRead` but the only way I could think of is unusable. It's not a big deal though.

There's still one case when I think `BufRead` might be better: If you call `fill_buf` and find out you have enough data to deserialize the whole type without entering any loops and doing offset computations then it might be more performant to do so. I have some ideas what to do about that too but that'll need more thought.

**Comment by dpc** [comment]
> BufRead requirement is paternalistic and impractical

Agreed. 💯
